system/camerad/cameras/spectra.cc

```diff
@@ -1384,67 +1336,102 @@ bool SpectraCamera::handle_camera_event(const cam_req_mgr_message *event_data) {
-        };
-        return true;
-      }
-      // LOGW("camerad %d synced req %d fid %d, publishing ts %.2f cereal_frame_id %d", cc.camera_num, (int)request_id, (int)frame_id_raw, (double)(timestamp)*1e-6, meta_data.frame_id);
-    } else {
-      // Frames not yet synced
-      enqueue_req_multi(request_id + ife_buf_depth, 1);
-      // LOGW("camerad %d not synced req %d fid %d", cc.camera_num, (int)request_id, (int)frame_id_raw);
-    }
-  } else { // not ready
+  if (request_id == 0) {  // not ready
     if (frame_id_raw > frame_id_raw_last + 10) {
       LOGE("camera %d reset after half second of no response", cc.camera_num);
-      clear_req_queue();
-      enqueue_req_multi(request_id_last + 1, ife_buf_depth);
+      clearAndRequeue(request_id_last + 1);
       frame_id_raw_last = frame_id_raw;
-      skipped_last = true;
     }
+    return false;
   }
 
-  return false;
+  // check for skipped_last frames
+  if (frame_id_raw > frame_id_raw_last + 1 && !skipped_last) {
+    LOGE("camera %d realign", cc.camera_num);
+    clearAndRequeue(request_id + 1);
+  } else if (frame_id_raw == frame_id_raw_last + 1) {
+    skipped_last = false;
+  }
+
+  // check for dropped requests
+  if (!skipped_last && request_id > request_id_last + 1) {
+    LOGE("camera %d dropped requests %ld %ld", cc.camera_num, request_id, request_id_last);
+    clearAndRequeue(request_id_last + 1);
+  }
+
+  frame_id_raw_last = frame_id_raw;
+  request_id_last = request_id;
+
+  return processFrame(request_id, frame_id_raw, timestamp);
+}
+
+void SpectraCamera::clearAndRequeue(uint64_t from_request_id) {
+  clear_req_queue();
+  enqueue_req_multi(from_request_id, ife_buf_depth);
+  skipped_last = true;
+}
+
+bool SpectraCamera::waitForFrameReady(int buf_idx, uint64_t request_id) {
+  if (!sync_objs_ife[buf_idx]) return false;
+
+  // wait for frame from ISP
+  // - in RAW_OUTPUT mode, this time is just the frame readout from the sensor
+  // - in IFE_PROCESSED mode, this time is both frame readout and image processing (~1ms)
+  auto waitForSync = [&](uint32_t sync_obj, int timeout_ms, const char *sync_type) {
+    struct cam_sync_wait sync_wait = {};
+    sync_wait.sync_obj = sync_obj;
+    sync_wait.timeout_ms = stress_test(sync_type) ? 1 : timeout_ms;
+    int ret = do_sync_control(m->cam_sync_fd, CAM_SYNC_WAIT, &sync_wait, sizeof(sync_wait));
+    if (ret != 0) {
+      LOGE("Failed to wait for %s: %d %d", sync_type, ret, sync_obj);
+    }
+    return ret == 0;
+  };
+
+  bool success = waitForSync(sync_objs_ife[buf_idx], 100, "IFE sync");
+  if (success && sync_objs_bps[buf_idx]) {
+    // BPS is typically 7ms
+    success = waitForSync(sync_objs_bps[buf_idx], 50, "BPS sync");
+  }
+
+  if (success) {
+    destroySyncObjectAt(buf_idx);
+    enqueue_buffer(buf_idx, request_id + ife_buf_depth);
+  } else {
+    // Reset queue on sync failure to prevent frame tearing
+    clearAndRequeue(request_id + 1);
```

Comment:
so this is pretty good, but can we further centralize the clear and en-queuing to a single line or function at least?